sequencer/src/persistence/sql.rs

```diff
@@ -1966,6 +1966,123 @@ impl SequencerPersistence for Persistence {
         Ok(())
     }
 
+    async fn migrate_stake_table_events(&self) -> anyhow::Result<()> {
+        let batch_size: i64 = 10_000;
+        let mut tx = self.db.read().await?;
+
+        let migration_status = query_as::<(bool, i64)>(
+            "SELECT completed, migrated_rows 
+         FROM epoch_migration 
+         WHERE table_name = 'stake_table_events'",
+        )
+        .fetch_optional(tx.as_mut())
+        .await?;
+
+        let mut offset = if let Some((completed, migrated_rows)) = migration_status {
+            if completed {
+                tracing::info!("Migration already completed for stake_table_events");
+                return Ok(());
+            }
+            tracing::info!(
+                "Resuming stake_table_events migration from offset={}",
+                migrated_rows
+            );
+            migrated_rows
+        } else {
+            tracing::info!("No existing migration entry for stake_table_events, starting from 0");
+            0
+        };
+
+        tracing::warn!("migrating stake_table_events…");
+
+        loop {
+            let mut rtx = self.db.read().await?;
+            let rows = query(
+                "SELECT l1_block, log_index, event 
+             FROM stake_table_events
+             ORDER BY l1_block, log_index
+             LIMIT $1 OFFSET $2",
+            )
+            .bind(batch_size)
+            .bind(offset)
+            .fetch_all(rtx.as_mut())
+            .await?;
+            drop(rtx);
+
+            if rows.is_empty() {
+                break;
+            }
+
+            let mut values = Vec::new();
+
+            for row in rows.iter() {
+                let l1_block: i64 = row.try_get("l1_block")?;
+                let log_index: i64 = row.try_get("log_index")?;
+                let event_value: serde_json::Value = row.try_get("event")?;
+
+                // deserialize legacy
+                let legacy_event: StakeTableEventLegacy =
+                    serde_json::from_value(event_value.clone())
+                        .context("Failed to deserialize legacy stake_table_event")?;
+
+                let new_event: StakeTableEvent = legacy_event.into();
+                let event_json = serde_json::to_value(new_event)?;
+
+                values.push((l1_block, log_index, event_json));
+            }
+
+            // Insert batch into new table
+            let mut query_builder: sqlx::QueryBuilder<Db> = sqlx::QueryBuilder::new(
+                "INSERT INTO stake_table_events (l1_block, log_index, event) ",
+            );
+
+            query_builder.push_values(values, |mut b, (l1_block, log_index, event)| {
+                b.push_bind(l1_block).push_bind(log_index).push_bind(event);
+            });
+            query_builder.push(
+                " ON CONFLICT (l1_block, log_index) 
```

Comment:
Why do we need this?

Reply:
we already have events in the stake table events table. The ON CONFLICT will overwrite those events with the new serialization